dhis-2/dhis-services/dhis-service-tracker/src/test/java/org/hisp/dhis/tracker/validation/hooks/EventReferralValidationHookTest.java

```diff
@@ -0,0 +1,254 @@
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.program.ProgramInstance;
+import org.hisp.dhis.program.ProgramStage;
+import org.hisp.dhis.program.ProgramType;
+import org.hisp.dhis.relationship.RelationshipType;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityType;
+import org.hisp.dhis.tracker.TrackerIdSchemeParams;
+import org.hisp.dhis.tracker.bundle.TrackerBundle;
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.domain.MetadataIdentifier;
+import org.hisp.dhis.tracker.domain.Relationship;
+import org.hisp.dhis.tracker.domain.RelationshipItem;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.ValidationErrorReporter;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import com.google.common.collect.Sets;
+
+/**
+ * @author rajazubair
+ */
+
+@ExtendWith( MockitoExtension.class )
+public class EventReferralValidationHookTest extends DhisConvenienceTest
+{
+    public static final String ORG_UNIT_UID = CodeGenerator.generateUid();
+
+    public static final String TEI_TYPE_UID = CodeGenerator.generateUid();
+
+    public static final String TEI_UID = CodeGenerator.generateUid();
+
+    public static final String PROGRAM_UID = CodeGenerator.generateUid();
+
+    public static final String PROGRAM_STAGE_UID = CodeGenerator.generateUid();
+
+    public static final String PROGRAM_STAGE_REFERRAL_UID = CodeGenerator.generateUid();
+
+    public static final String ENROLLMENT_UID = CodeGenerator.generateUid();
+
+    private OrganisationUnit orgUnit;
+
+    private TrackedEntityType teiType;
+
+    private Program program;
+
+    private ProgramStage programStage;
+
+    private ProgramStage programStageReferral;
+
+    private TrackedEntityInstance tei;
+
+    private ProgramInstance programInstance;
+
+    private Relationship relationship;
+
+    private Relationship relationshipIncomplete;
+
+    private EventReferralValidationHook subject;
+
+    private TrackerBundle bundle;
+
+    @Mock
+    private TrackerPreheat preheat;
+
+    private ValidationErrorReporter reporter;
+
+    @BeforeEach
+    void setUp()
+    {
+        subject = new EventReferralValidationHook();
+
+        bundle = TrackerBundle.builder().preheat( preheat ).build();
+        TrackerIdSchemeParams idSchemes = TrackerIdSchemeParams.builder().build();
+        reporter = new ValidationErrorReporter( idSchemes );
+
+        init();
+    }
+
+    @Test
+    public void verifyValidationSuccessForNonReferralEvent()
+    {
+        when( preheat.getProgramStage( MetadataIdentifier.ofUid( PROGRAM_STAGE_UID ) ) )
+            .thenReturn( programStage );
+
+        Event event = Event.builder()
```

Comment:
You have events that you need for the test that will only differ by programStage and relationship. You can a private method that will get this 2 parameter and create the event to test.
This way is good to clean up the test and remove the focus on building object details.

Reply:
resolved